atom/src/space.rs

```diff
@@ -0,0 +1,480 @@
+//! Smart pointers with safe atom reading and writing methods.
+//!
+//! # Safety
+//!
+//! The only unsafe things that happen in this module is when either space is created from a reference to a `sys::LV2_Atom` and when space is re-interpreted as typed data.
+//!
+//! In the first case, we have to trust that the space behind the atom header is accessible since we have no way to check whether it is or not. Therefore, we have to assume that it is sound.
+//!
+//! The second case is sound since a) the data is contained in a slice and therefore is accessible, b) generic type parameter bounds assure that the type is plain-old-data and c) 64-bit padding is assured.
+use std::cell::Cell;
+use std::marker::Unpin;
+use std::mem::{size_of, size_of_val};
+use urid::URID;
+
+/// Specialized smart pointer to retrieve struct instances from a slice of memory.
+///
+/// The accessor methods of this struct all behave in a similar way: If the internal slice is big enough, they create a reference to the start of the slice with the desired type and create a new space object that contains the space after the references instance.
+#[derive(Clone, Copy)]
+pub struct Space<'a> {
+    data: Option<&'a [u8]>,
+}
+
+impl<'a> Space<'a> {
+    /// Create a new space from an atom pointer.
+    ///
+    /// The method creates a space that contains the atom as well as it's body. Since the body is not included in the atom reference, this method has to assume that it is valid memory and therefore is unsafe but sound.
+    #[allow(clippy::trivially_copy_pass_by_ref)]
```

Comment:
Yes, `sys::LV2_Atom` is `Copy`, because the type itself is qualified to be: It can be duplicated by copying the bits from one place to another. Its presence does not necessarily imply the presence of a buffer behind it. This is only an interpretation done in `Space::from_atom`. Any other methods imply the size of the buffer behind the atom from other sources.

On the other hand, I had to restrict value the `MutSpace::write` method to types that are `Copy`, because it copies the value byte per byte from one place to another. This means that `sys::LV2_Atom` has to implement `Copy`.

Reply:
Oh i see, fair enough! :slightly_smiling_face: